gammapy/irf/energy_dispersion.py

```diff
@@ -726,65 +728,27 @@ def get_response(self, offset, e_true, e_reco=None):
-            center=e_reco.log_centers
-            migra=center / e_true
+        # oversample migra
+        migra_lo = migra[:-1]
+        migra_hi = migra[1:]
+        #migra_range = np.column_stack([migra_lo, migra_hi]])
+        migra_grid = list()
 
-        # ensure to have a normalized edisp
-        migra_bin=self.migra.hi - self.migra.lo
-        if (migra_bin[0] == migra_bin[1]):
-            migra_mean=1 / 2. * (self.migra.hi + self.migra.lo)
-        else:
-            migra_mean=np.sqrt(self.migra.hi * self.migra.lo)
-        val_norm=self.data.evaluate(offset=offset, e_true=e_true, migra=migra_mean)
-        norm=np.sum(val_norm * migra_bin)
-
-        migra_e_reco=e_reco / e_true
-        integral=np.zeros(len(e_reco) - 1)
-        if norm != 0:
-            for i, migra_int in enumerate(migra_e_reco[:-1]):
-                # The migra_e_reco bin is inferior to the migra min in the dispersion fits file
-                if migra_e_reco[i + 1] < migra_mean[0]:
-                    continue
-                # The migra_e_reco bin is superior to the migra max in the dispersion fits file
-                elif migra_e_reco[i] > migra_mean[-1]:
-                    continue
-                else:
-                    if migra_e_reco[i] < migra_mean[0]:
-                        i_min=0
-                    else:
-                        i_min=np.where(migra_mean < migra_e_reco[i])[0][-1]
-                    i_max=np.where(migra_mean < migra_e_reco[i + 1])[0][-1]
-
-                    migra_bin_reco=migra_bin[i_min + 1:i_max]
-                    if migra_e_reco[i + 1] > migra_mean[-1]:
-                        index=np.arange(i_min, i_max, 1)
-                        migra_bin_reco=np.insert(migra_bin_reco, 0, (migra_mean[i_min + 1] - migra_e_reco[i]))
-                    elif migra_e_reco[i] < migra_mean[0]:
-                        if i_max == 0:
-                            index=np.arange(i_min, i_max + 1, 1)
-                        else:
-                            index=np.arange(i_min + 1, i_max + 1, 1)
-                        migra_bin_reco=np.append(migra_bin_reco, migra_e_reco[i + 1] - migra_mean[i_max])
-                    elif i_min == i_max:
-                        index=np.arange(i_min, i_max + 1, 1)
-                        migra_bin_reco=np.insert(migra_bin_reco, 0, (migra_e_reco[i + 1] - migra_e_reco[i]))
-                    else:
-                        index=np.arange(i_min, i_max + 1, 1)
-                        migra_bin_reco=np.insert(migra_bin_reco, 0, (migra_mean[i_min + 1] - migra_e_reco[i]))
-                        migra_bin_reco=np.append(migra_bin_reco, migra_e_reco[i + 1] - migra_mean[i_max])
-                    val=self.data.evaluate(offset=offset, e_true=e_true, migra=migra_mean[index])
-                    integral[i]=np.sum(val * migra_bin_reco) / norm
-        return integral
+        # TODO: Get rid of this loop
+        for mlo, mhi in zip(migra_lo, migra_hi):
```

Comment:
> I thinks it's annoying to have to vectorize everything.

Well, we have 1000s of lines of code where it's simpler than if we had to write for loops everywhere. But yes, I agree, it can be tricky and annoying.
Somehow it's surprising that in Gammapy we almost never encounter difficult cases, probably because our data is so simple (always 1-dim or 2-dim or 3-dim arrays and simple operations).
The spectrum energy grouping is a difficult case, but there we decided to just use Python data structures and not try to vectorise, because efficency is not a concern to compute the energy binning.